src/main/java/org/jenkinsci/plugins/durabletask/WindowsBatchScript.java

```diff
@@ -54,36 +61,171 @@ public String getScript() {
         BatchController c = new BatchController(ws);
 
-        c.getBatchFile1(ws).write(String.format("cmd /c \"\"%s\"\" > \"%s\" 2>&1\r\necho %%ERRORLEVEL%% > \"%s\"\r\n",
-                c.getBatchFile2(ws).getRemote().replace("%", "%%"),
-                c.getLogFile(ws).getRemote().replace("%", "%%"),
-                c.getResultFile(ws).getRemote().replace("%", "%%")
-        ), "UTF-8");
-        c.getBatchFile2(ws).write(script, "UTF-8");
+        /*
+          This wrapper script basically only calls the actual script that is supposed to be executed. It redirects all
+          the output of the target script to the log file and stores the exit code in a result file.
+          The script uses a additional temporary to store the exit code that is moved after writing to is is done.
+          Writing the output file directly may cause problems because the check for the exit code may locate the result
+          file after it was created but before the code was written in and issue a error because of this.
+         */
+        final String nl = "\r\n";
+        final String identifier = c.getIdentifier();
 
-        Launcher.ProcStarter ps = launcher.launch().cmds("cmd", "/c", "\"\"" + c.getBatchFile1(ws) + "\"\"").envs(envVars).pwd(ws).quiet(true);
+        c.getWrapperBatchFile(ws).write(
+                "CHCP 65001 > NUL" + nl +
+                "SETLOCAL" + nl +
+                "SET theRealScript=" + escapeForBatch(c.getMainBatchFile(ws)) + nl +
+                "SET resultFile=" + escapeForBatch(c.getResultFile(ws)) + nl +
+                "SET logFile=" + escapeForBatch(c.getLogFile(ws)) + nl +
+                "SET tempResultFile=" + escapeForBatch(c.getTemporaryResultFile(ws)) + nl +
+                "SET pidFile=" + escapeForBatch(c.getPidFile(ws)) + nl +
+                "TITLE=" + identifier + nl + /* This works regardless if the command windows is actually visible. */
+                "FOR /F \"tokens=*\" %%1 IN ('tasklist /NH /FI \"IMAGENAME eq cmd.exe\" /V /FO CSV') DO (" + nl +
+                "  ECHO %%1 | FINDSTR /C:" + identifier + " >NUL 2>&1 && (" + nl +
+                "    FOR /F \"tokens=2 delims=,\" %%P IN (\"%%1\") DO ( " + nl +
+                "      ECHO %%~P > \"%pidFile%\"" + nl +
+                "    )" + nl +
+                "  )" + nl +
+                ")" + nl +
+                "TYPE NUL > \"%tempResultFile%\"" + nl +
+                "CMD /C \"%theRealScript%\" > \"%logFile%\" 2>&1" + nl +
+                "ECHO %ERRORLEVEL% > \"%tempResultFile%\"" + nl +
+                "MOVE \"%tempResultFile%\" \"%resultFile%\" > NUL" + nl +
+                "DEL /Q %pidFile% > NUL 2>&1" + nl +
+                "ENDLOCAL" + nl +
+                "EXIT 0", "UTF-8");
+        c.getMainBatchFile(ws).write(script, "UTF-8");
+
+        Launcher.ProcStarter ps = launcher.launch()
+                                          .cmds("cmd", "/C", "START \"\" /MIN \"" + escapeForBatch(c.getWrapperBatchFile(ws)) + '"')
```

Comment:
I guess there should be double quotes \"\" to handle paths with spaces correctly (see JENKINS-25678)